src/Frame.cpp

```diff
@@ -1,251 +1,626 @@
-    {"O", 1.3},
-    {"N", 1.4},
-    {"S", 1.9},
-    {"F", 1.2},
+	{"H", 1.0},
+	{"C", 1.5},
+	{"O", 1.3},
+	{"N", 1.4},
+	{"S", 1.9},
+	{"F", 1.2},
 };
 
 optional<double> guess_bonds_radius(const Atom& atom) {
-    const auto& type = atom.type();
-    auto it = BOND_GUESSING_RADII.find(type);
-    if (it != BOND_GUESSING_RADII.end()) {
-        // allow configuration file to override data from BOND_GUESSING_RADII
-        auto user_config = Configuration::atom_data(type);
-        if (user_config && user_config->vdw_radius) {
-            return user_config->vdw_radius.value();
-        } else {
-            return it->second;
-        }
-    } else {
-        // default to chemfiles provided atom type
-        return atom.vdw_radius();
-    }
+	const auto& type = atom.type();
+	auto it = BOND_GUESSING_RADII.find(type);
+	if (it != BOND_GUESSING_RADII.end()) {
+		// allow configuration file to override data from BOND_GUESSING_RADII
+		auto user_config = Configuration::atom_data(type);
+		if (user_config && user_config->vdw_radius) {
+			return user_config->vdw_radius.value();
+		}
+		else {
+			return it->second;
+		}
+	}
+	else {
+		// default to chemfiles provided atom type
+		return atom.vdw_radius();
+	}
+}
+
+int Frame::generateBonds(std::vector<int>* result, std::vector<Vector3D>* positions, int totalCells, int** cells,
+	int* numInCell, int** neighborList, int maxBonds, float cutoff)
+{
+	int i, j, cellIndex;
+	int bondsCount = 0;
+
+	float cutoffPow2 = cutoff * cutoff;
+
+	for (cellIndex = 0; (cellIndex < totalCells) && (bondsCount < maxBonds); cellIndex += 1) {
+
+		const int* tmpCell = cells[cellIndex];
+		int atomsInCell = numInCell[cellIndex];
+
+		if (((cellIndex & 255) == 0)) {
+			char tmpbuf[128];
+			sprintf(tmpbuf, "%6.2f", (100.0f * cellIndex) / (float)totalCells);
+			printf("Generating bonds...: %s %% complete\n", tmpbuf);
+		}
+
+		for (const int* neighborCellsIndexes = neighborList[cellIndex]; (*neighborCellsIndexes != -1) && (bondsCount < maxBonds); neighborCellsIndexes++) {
+
+			int neighborCellIndex = *neighborCellsIndexes;
+			const int* neighborCell = cells[neighborCellIndex];
+			int atomNumInCell = numInCell[neighborCellIndex];
+			bool selfCell;
+			if (cellIndex == neighborCellIndex)
+				selfCell = true;
+			else
+				selfCell = false;
+
+			for (i = 0; (i < atomsInCell) && (bondsCount < maxBonds); i++) {
+				int atom1 = tmpCell[i];
+				Vector3D posAtom1 = positions->at(atom1);
+
+				//Skip over self and already-tested atoms
+				int firstNeighborCell = (selfCell) ? i + 1 : 0;
+
+				for (j = firstNeighborCell; (j < atomNumInCell) && (bondsCount < maxBonds); j++) {
+					int atom2 = neighborCell[j];
+					Vector3D posAtom2 = positions->at(atom2);
+					float distanceX = posAtom1[0] - posAtom2[0];
+					float distanceY = posAtom1[1] - posAtom2[1];
+					float distanceZ = posAtom1[2] - posAtom2[2];
+					float interatomicDistance = distanceX * distanceX + distanceY * distanceY + distanceZ * distanceZ;
```

Comment:
This does not account for periodic boundary conditions. Why not use `this->distance(atom1, atom2)` here?

Reply:
You are right, it clearly doesn't take into account PBC but does bond perception have to take PBC into account? If yes, it means that when one of a two bonded atom crosses the box limit and goes to the other side of the box, the bond will be displayed in the entire width of the box. I don't know how PBC are treated and displayed in chemfiles but I could definitely use Frame's distance function in order to take that into account.

Reply:
Yes, I think that bond perception should take pbc into account, because it is not used only for visualization. For example I used this code to generate LAMMPS input from XYZ files, and in this case we want to get all bonds in the system, even the ones crossing periodic boundary.

To prevent this code from generating ugly bonds in a visualization application, I could see two approaches: one is to set the cell to an infinite cell before this calculation, and reset the original cell afterward. This would make the distance calculation behave exactly as the code as currently written. The other approach is the one taken by [VESTA](http://jp-minerals.org/vesta/en/), where broken molecules get duplicated when a bond would cross the cell. See the image below, where the input only contains atoms inside the cell
![paracetamol](https://user-images.githubusercontent.com/6141587/98549471-1f93fe80-229b-11eb-870a-bd2caf288888.png)

Reply:
Thank you, your input is interesting and it makes sense. I will modify the code accordingly and prefer `this->distance(atom1,` atom2)` instead of a standard distance calculation.